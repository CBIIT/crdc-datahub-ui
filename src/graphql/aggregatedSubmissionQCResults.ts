import gql from "graphql-tag";

// The base aggregatedQCResult model used for all aggregatedSubmissionQCResults queries
const BaseAggregatedQCResultFragment = gql`
  fragment BaseAggregatedQCResultFragment on aggregatedQCResult {
    code
    title
  }
`;

// The extended aggregatedQCResult model which includes all fields
const FullAggregatedQCResultFragment = gql`
  fragment AggregatedQCResultFragment on aggregatedQCResult {
    severity
    count
  }
`;

export const query = gql`
  query aggregatedSubmissionQCResults(
    $submissionID: ID!
    $severity: String
    $first: Int
    $offset: Int
    $orderBy: String
    $sortDirection: String
    $partial: Boolean = false
  ) {
    aggregatedSubmissionQCResults(
      submissionID: $submissionID
      severity: $severity
      first: $first
      offset: $offset
      orderBy: $orderBy
      sortDirection: $sortDirection
    ) {
      total
      results {
        ...BaseAggregatedQCResultFragment
        ...AggregatedQCResultFragment @skip(if: $partial)
      }
    }
  }
  ${FullAggregatedQCResultFragment}
  ${BaseAggregatedQCResultFragment}
`;

export type Input = {
  submissionID: string;
  severity?: string;
  first?: number;
  offset?: number;
  orderBy?: string;
  sortDirection?: string;
  partial?: boolean;
};

export type Response = {
  aggregatedSubmissionQCResults: ValidationResult<AggregatedQCResult>;
};
