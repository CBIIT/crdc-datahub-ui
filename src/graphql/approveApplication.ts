import gql from "graphql-tag";

export const mutation = gql`
  mutation approveApplication(
    $id: ID!
    $comment: String
    $wholeProgram: Boolean
    $institutions: [String]
    $pendingModelChange: Boolean
  ) {
    approveApplication(
      _id: $id
      wholeProgram: $wholeProgram
      comment: $comment
      institutions: $institutions
      pendingModelChange: $pendingModelChange
    ) {
      _id
    }
  }
`;

export type Input = {
  id: string;
  comment: string;
  wholeProgram: boolean;
  institutions: string[];
  pendingModelChange: boolean;
};

export type Response = {
  approveApplication: Pick<Application, "_id">;
};
