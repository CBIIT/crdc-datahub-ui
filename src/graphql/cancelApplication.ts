import { TypedDocumentNode } from "@apollo/client";
import gql from "graphql-tag";

export const mutation: TypedDocumentNode<Response, Input> = gql`
  mutation cancelApplication($_id: ID!) {
    cancelApplication: deleteApplication(_id: $_id) {
      _id
    }
  }
`;

export type Input = {
  /**
   * ID of the application to cancel
   */
  _id: string;
};

export type Response = {
  cancelApplication: Pick<Application, "_id">;
};
