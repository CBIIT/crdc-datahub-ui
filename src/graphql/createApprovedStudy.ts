import gql from "graphql-tag";

export const mutation = gql`
  mutation createApprovedStudy(
    $name: String!
    $acronym: String
    $controlledAccess: Boolean!
    $openAccess: Boolean
    $dbGaPID: String
    $ORCID: String
    $PI: String
    $primaryContactID: String
  ) {
    createApprovedStudy(
      name: $name
      acronym: $acronym
      controlledAccess: $controlledAccess
      openAccess: $openAccess
      dbGaPID: $dbGaPID
      ORCID: $ORCID
      PI: $PI
      primaryContactID: $primaryContactID
    ) {
      _id
      studyName
      studyAbbreviation
      dbGaPID
      controlledAccess
      openAccess
      PI
      ORCID
      programs {
        _id
      }
      createdAt
    }
  }
`;

export type Input = {
  name: string;
  acronym: string;
  controlledAccess: boolean;
  openAccess: boolean;
  dbGaPID: string;
  ORCID: string;
  PI: string;
  primaryContactID: string;
};

export type Response = {
  createApprovedStudy: ApprovedStudy;
};
