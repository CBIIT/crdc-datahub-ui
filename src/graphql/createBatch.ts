import gql from "graphql-tag";

export const mutation = gql`
  mutation createBatch($submissionID: ID!, $type: String, $files: [String!]!) {
    createBatch(submissionID: $submissionID, type: $type, files: $files) {
      _id
      submissionID
      bucketName
      filePrefix
      type
      fileCount
      files {
        fileName
        signedURL
      }
      status
      createdAt
      updatedAt
    }
  }
`;

export type Input = {
  submissionID: string;
  type: UploadType;
  files: string[];
};

export type Response = {
  createBatch: NewBatch;
};
