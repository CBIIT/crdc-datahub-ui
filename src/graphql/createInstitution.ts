import { TypedDocumentNode } from "@apollo/client";
import gql from "graphql-tag";

export const mutation: TypedDocumentNode<Response, Input> = gql`
  mutation createInstitution($name: String!, $status: String!) {
    createInstitution(name: $name, status: $status) {
      _id
    }
  }
`;

export type Input = {
  name: string;
  status: string;
};

export type Response = {
  createInstitution: Pick<Institution, "_id">;
};
