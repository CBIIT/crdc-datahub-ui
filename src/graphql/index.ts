// Submission Request
export { mutation as REVIEW_APP } from "./reviewApplication";
export type { Response as ReviewAppResp, Input as ReviewAppInput } from "./reviewApplication";

export { mutation as REOPEN_APP } from "./reopenApplication";
export type { Response as ReopenAppResp } from "./reopenApplication";

export { mutation as APPROVE_APP } from "./approveApplication";
export type { Input as ApproveAppInput, Response as ApproveAppResp } from "./approveApplication";

export { mutation as INQUIRE_APP } from "./inquireApplication";
export type { Response as InquireAppResp } from "./inquireApplication";

export { mutation as REJECT_APP } from "./rejectApplication";
export type { Response as RejectAppResp } from "./rejectApplication";

export { mutation as CANCEL_APP } from "./cancelApplication";
export type { Response as CancelAppResp, Input as CancelAppInput } from "./cancelApplication";

export { mutation as RESTORE_APP } from "./restoreApplication";
export type { Response as RestoreAppResp, Input as RestoreAppInput } from "./restoreApplication";

export { mutation as SAVE_APP } from "./saveApplication";
export type { Input as SaveAppInput, Response as SaveAppResp } from "./saveApplication";

export { mutation as SUBMIT_APP } from "./submitApplication";
export type { Response as SubmitAppResp } from "./submitApplication";

export { query as LAST_APP } from "./getMyLastApplication";
export type { Response as LastAppResp } from "./getMyLastApplication";

export { query as GET_APP } from "./getApplication";
export type { Response as GetAppResp } from "./getApplication";

export { mutation as UPDATE_MY_USER } from "./updateMyUser";
export type { Input as UpdateMyUserInput, Response as UpdateMyUserResp } from "./updateMyUser";

export { query as LIST_INSTITUTIONS } from "./listInstitutions";
export type { Response as ListInstitutionsResp } from "./listInstitutions";

export { query as RETRIEVE_CDEs } from "./retrieveCDEs";
export type { Response as RetrieveCDEsResp, Input as RetrieveCDEsInput } from "./retrieveCDEs";

export { query as LIST_APPLICATIONS } from "./listApplications";
export type {
  Response as ListApplicationsResp,
  Input as ListApplicationsInput,
} from "./listApplications";

// Data Submissions
export { mutation as CREATE_SUBMISSION } from "./createSubmission";
export type {
  Input as CreateSubmissionInput,
  Response as CreateSubmissionResp,
} from "./createSubmission";

export { query as GET_SUBMISSION } from "./getSubmission";
export type { Input as GetSubmissionInput, Response as GetSubmissionResp } from "./getSubmission";

export { query as LIST_SUBMISSIONS } from "./listSubmissions";
export type {
  Input as ListSubmissionsInput,
  Response as ListSubmissionsResp,
} from "./listSubmissions";

export { mutation as SUBMISSION_ACTION } from "./submissionAction";
export type { Response as SubmissionActionResp } from "./submissionAction";

export { mutation as CREATE_BATCH } from "./createBatch";
export type { Input as CreateBatchInput, Response as CreateBatchResp } from "./createBatch";

export { mutation as UPDATE_BATCH } from "./updateBatch";
export type { Response as UpdateBatchResp } from "./updateBatch";

export { query as LIST_BATCHES } from "./listBatches";
export type { Input as ListBatchesInput, Response as ListBatchesResp } from "./listBatches";

export { query as SUBMISSION_QC_RESULTS } from "./submissionQCResults";
export type {
  Input as SubmissionQCResultsInput,
  Response as SubmissionQCResultsResp,
} from "./submissionQCResults";

export { query as AGGREGATED_SUBMISSION_QC_RESULTS } from "./aggregatedSubmissionQCResults";
export type {
  Input as AggregatedSubmissionQCResultsInput,
  Response as AggregatedSubmissionQCResultsResp,
} from "./aggregatedSubmissionQCResults";

export { query as SUBMISSION_CROSS_VALIDATION_RESULTS } from "./submissionCrossValidationResults";
export type {
  Input as CrossValidationResultsInput,
  Response as CrossValidationResultsResp,
} from "./submissionCrossValidationResults";

export { mutation as VALIDATE_SUBMISSION } from "./validateSubmission";
export type {
  Input as ValidateSubmissionInput,
  Response as ValidateSubmissionResp,
} from "./validateSubmission";

export { query as LIST_NODE_TYPES } from "./listSubmissionNodeTypes";
export type {
  Input as ListNodeTypesInput,
  Response as ListNodeTypesResp,
} from "./listSubmissionNodeTypes";

export { query as GET_NODE_DETAIL } from "./getNodeDetail";
export type { Input as GetNodeDetailInput, Response as GetNodeDetailResp } from "./getNodeDetail";

export { query as GET_RELATED_NODES } from "./getRelatedNodes";
export type {
  Input as GetRelatedNodesInput,
  Response as GetRelatedNodesResp,
  PropertiesOnlyResponse as GetRelatedNodesRespPropsOnly,
} from "./getRelatedNodes";

export { query as GET_SUBMISSION_NODES } from "./getSubmissionNodes";
export type {
  Input as GetSubmissionNodesInput,
  Response as GetSubmissionNodesResp,
} from "./getSubmissionNodes";

export { query as RETRIEVE_RELEASED_DATA } from "./retrieveReleasedDataByID";
export type {
  Input as RetrieveReleasedDataInput,
  Response as RetrieveReleasedDataResp,
} from "./retrieveReleasedDataByID";

export { query as SUBMISSION_STATS } from "./submissionStats";
export type {
  Input as SubmissionStatsInput,
  Response as SubmissionStatsResp,
} from "./submissionStats";

export { mutation as DELETE_DATA_RECORDS } from "./deleteDataRecords";
export type {
  Input as DeleteDataRecordsInput,
  Response as DeleteDataRecordsResp,
} from "./deleteDataRecords";

export { query as RETRIEVE_CLI_CONFIG } from "./retrieveCLIConfig";
export type { Response as RetrieveCLIConfigResp } from "./retrieveCLIConfig";

export { query as LIST_POTENTIAL_COLLABORATORS } from "./listPotentialCollaborators";
export type {
  Input as ListPotentialCollaboratorsInput,
  Response as ListPotentialCollaboratorsResp,
} from "./listPotentialCollaborators";

export { mutation as EDIT_SUBMISSION_COLLABORATORS } from "./editSubmissionCollaborators";
export type {
  Input as EditSubmissionCollaboratorsInput,
  Response as EditSubmissionCollaboratorsResp,
} from "./editSubmissionCollaborators";

// User Profile
export { query as GET_MY_USER } from "./getMyUser";
export type { Response as GetMyUserResp } from "./getMyUser";

export { query as GET_USER } from "./getUser";
export type { Input as GetUserInput, Response as GetUserResp } from "./getUser";

export { query as LIST_USERS } from "./listUsers";
export type { Response as ListUsersResp } from "./listUsers";

export { mutation as EDIT_USER } from "./editUser";
export type { Input as EditUserInput, Response as EditUserResp } from "./editUser";

export { mutation as REQUEST_ACCESS } from "./requestAccess";
export type { Input as RequestAccessInput, Response as RequestAccessResp } from "./requestAccess";

export { query as RETRIEVE_PBAC_DEFAULTS } from "./retrievePBACDefaults";
export type {
  Input as RetrievePBACDefaultsInput,
  Response as RetrievePBACDefaultsResp,
} from "./retrievePBACDefaults";

// Organizations
export { query as LIST_ORGS } from "./listOrganizations";
export type { Response as ListOrgsResp } from "./listOrganizations";

export { query as GET_ORG } from "./getOrganization";
export type { Response as GetOrgResp } from "./getOrganization";

export { mutation as EDIT_ORG } from "./editOrganization";
export type { Input as EditOrgInput, Response as EditOrgResp } from "./editOrganization";

export { query as LIST_ACTIVE_DCPS } from "./listActiveDCPs";
export type { Response as ListActiveDCPsResp } from "./listActiveDCPs";

export { query as LIST_APPROVED_STUDIES } from "./listApprovedStudies";
export type {
  Input as ListApprovedStudiesInput,
  Response as ListApprovedStudiesResp,
} from "./listApprovedStudies";

export { mutation as CREATE_APPROVED_STUDY } from "./createApprovedStudy";
export type {
  Input as CreateApprovedStudyInput,
  Response as CreateApprovedStudyResp,
} from "./createApprovedStudy";

export { mutation as UPDATE_APPROVED_STUDY } from "./updateApprovedStudy";
export type {
  Input as UpdateApprovedStudyInput,
  Response as UpdateApprovedStudyResp,
} from "./updateApprovedStudy";

export { query as GET_APPROVED_STUDY } from "./getApprovedStudy";
export type {
  Input as GetApprovedStudyInput,
  Response as GetApprovedStudyResp,
} from "./getApprovedStudy";

export { mutation as CREATE_ORG } from "./createOrganization";
export type { Input as CreateOrgInput, Response as CreateOrgResp } from "./createOrganization";

// Misc.
export { mutation as GRANT_TOKEN } from "./grantToken";
export type { Response as GrantTokenResp } from "./grantToken";

// Operation Dashboard
export { query as GET_DASHBOARD_URL } from "./getDashboardURL";
export type {
  Input as GetDashboardURLInput,
  Response as GetDashboardURLResp,
} from "./getDashboardURL";
