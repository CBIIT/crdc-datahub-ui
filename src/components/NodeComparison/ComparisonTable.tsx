import {
  Alert,
  Skeleton,
  styled,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";
import { FC, memo, useMemo } from "react";
import { RetrieveReleasedDataResp } from "../../graphql";
import { safeParse } from "../../utils";
import Repeater from "../Repeater";

const StyledAlert = styled(Alert)({
  marginTop: "12px",
});

const StyledTableContainer = styled(TableContainer)({
  marginTop: "12px",
  paddingBottom: "6px",
  overflowX: "auto",
  width: "100%",
});

const StyledTable = styled(Table)({
  backgroundColor: "#fff",
  borderCollapse: "separate",
  borderRadius: "8px",
  tableLayout: "auto",
  // NOTE: The first element in this thead is a TD, not a TH
  "& thead tr:first-of-type td:first-of-type": {
    borderTopLeftRadius: "8px",
  },
  "& thead tr:first-of-type th:last-child": {
    borderTopRightRadius: "8px",
  },
  "& tbody tr:last-child td:first-of-type": {
    borderBottomLeftRadius: "8px",
  },
  "& tbody tr:last-child td:last-child": {
    borderBottomRightRadius: "8px",
  },
});

const StyledTableHead = styled(TableHead)({
  "& .MuiTableRow-root th": {
    fontWeight: 700,
  },
});

const StyledTableCell = styled(TableCell)({
  color: "#083A50",
  whiteSpace: "nowrap",
  overflow: "hidden",
  border: "0.5px solid #a7a7a7",
  padding: "8px 15px",
});

/**
 * The number of placeholder columns to display in the loading state
 */
const PLACEHOLDER_NUM_COLS = 5;

/**
 * The width of the new/existing column
 */
const BLANK_COL_WIDTH = 55;

export type ComparisonTableProps = {
  /**
   * The new node to be compared
   */
  newNode: RetrieveReleasedDataResp["retrieveReleasedDataByID"][number];
  /**
   * The existing node to be compared
   */
  existingNode: RetrieveReleasedDataResp["retrieveReleasedDataByID"][number];
  /**
   * A boolean indicating whether the data is in a loading state
   */
  loading: boolean;
};

/**
 * Builds a dynamic table to compare the dataset between two data records
 * Handles loading state
 *
 * @returns The NodeComparisonTable component
 */
const ComparisonTable: FC<ComparisonTableProps> = ({ newNode, existingNode, loading }) => {
  const newProps = useMemo<Record<string, string | number>>(
    () => safeParse(newNode?.props),
    [newNode]
  );

  const existingProps = useMemo<Record<string, string | number>>(
    () => safeParse(existingNode?.props),
    [existingNode]
  );

  const allPropertyNames = useMemo<string[]>(() => {
    const newKeys = Object.keys(newProps);
    const existingKeys = Object.keys(existingProps);
    return [...new Set([...newKeys, ...existingKeys])];
  }, [newProps, existingProps]);

  if (!loading && !allPropertyNames.length) {
    return (
      <StyledAlert severity="warning" data-testid="node-comparison-error">
        Unable to show comparison of data.
      </StyledAlert>
    );
  }

  return (
    <StyledTableContainer>
      <StyledTable data-testid="node-comparison-table" size="small">
        <StyledTableHead>
          <TableRow>
            <StyledTableCell component="td" width={BLANK_COL_WIDTH} />
            {loading ? (
              <TableHeaderSkeleton />
            ) : (
              allPropertyNames.map((property) => (
                <StyledTableCell key={property}>{property}</StyledTableCell>
              ))
            )}
          </TableRow>
        </StyledTableHead>
        <TableBody>
          {loading ? (
            <TableBodySkeleton />
          ) : (
            <>
              <TableRow data-testid="node-comparison-table-existing">
                <StyledTableCell width={BLANK_COL_WIDTH}>Existing</StyledTableCell>
                {allPropertyNames.map((property) => (
                  <StyledTableCell key={property}>
                    {existingProps?.[property] || ""}
                  </StyledTableCell>
                ))}
              </TableRow>
              <TableRow data-testid="node-comparison-table-new">
                <StyledTableCell width={BLANK_COL_WIDTH}>New</StyledTableCell>
                {allPropertyNames.map((property) => (
                  <StyledTableCell key={property}>{newProps?.[property] || ""}</StyledTableCell>
                ))}
              </TableRow>
            </>
          )}
        </TableBody>
      </StyledTable>
    </StyledTableContainer>
  );
};

/**
 * A skeleton table header to display while the data is loading
 *
 * @returns The TableHeaderSkeleton component
 */
const TableHeaderSkeleton: FC = () => (
  <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-header">
    <StyledTableCell component="td" data-testid="node-comparison-table-header-skeleton">
      <Skeleton variant="text" width="100%" height={24} />
    </StyledTableCell>
  </Repeater>
);

/**
 * A skeleton table body to display while the data is loading
 *
 * @returns The TableBodySkeleton component
 */
const TableBodySkeleton: FC = () => (
  <>
    <TableRow>
      <StyledTableCell width={BLANK_COL_WIDTH}>Existing</StyledTableCell>
      <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-existing">
        <StyledTableCell>
          <Skeleton variant="text" width="100%" height={24} />
        </StyledTableCell>
      </Repeater>
    </TableRow>
    <TableRow>
      <StyledTableCell width={BLANK_COL_WIDTH}>New</StyledTableCell>
      <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-new">
        <StyledTableCell>
          <Skeleton variant="text" width="100%" height={24} />
        </StyledTableCell>
      </Repeater>
    </TableRow>
  </>
);

export default memo<ComparisonTableProps>(ComparisonTable);
