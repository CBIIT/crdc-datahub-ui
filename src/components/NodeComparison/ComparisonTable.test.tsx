import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import ComparisonTable from "./ComparisonTable";
import { RetrieveReleasedDataResp } from "../../graphql";

const baseNode: RetrieveReleasedDataResp["retrieveReleasedDataByID"][number] = {
  nodeType: "",
  nodeID: "",
  props: "{}",
};

describe("Accessibility", () => {
  it("should have no violations", async () => {
    const { container, getByText } = render(
      <ComparisonTable
        newNode={{ ...baseNode, props: JSON.stringify({ mock_node_data_name: "bar", baz: 1 }) }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "baz", baz: 2 }),
        }}
        loading={false}
      />
    );

    expect(getByText(/mock_node_data_name/i)).toBeInTheDocument();

    expect(await axe(container)).toHaveNoViolations();
  });

  it("should have no violations (loading)", async () => {
    const { container } = render(<ComparisonTable newNode={null} existingNode={null} loading />);

    expect(await axe(container)).toHaveNoViolations();
  });

  it("should have no violations (no data)", async () => {
    const { container } = render(
      <ComparisonTable newNode={null} existingNode={null} loading={false} />
    );

    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Basic Functionality", () => {
  it("should render without crashing", () => {
    expect(() =>
      render(<ComparisonTable newNode={null} existingNode={null} loading={false} />)
    ).not.toThrow();
  });

  it("should show an error message when there is no data (all null)", () => {
    const { getByTestId, getByText } = render(
      <ComparisonTable newNode={null} existingNode={null} loading={false} />
    );

    expect(getByTestId(/node-comparison-error/i)).toBeInTheDocument();
    expect(getByText(/Unable to show comparison of data/i)).toBeInTheDocument();
  });

  it("should show an error message when there is no data (empty props)", () => {
    const { getByTestId, getByText } = render(
      <ComparisonTable
        newNode={{ ...baseNode, props: "{}" }}
        existingNode={{ ...baseNode, props: "{}" }}
        loading={false}
      />
    );

    expect(getByTestId(/node-comparison-error/i)).toBeInTheDocument();
    expect(getByText(/Unable to show comparison of data/i)).toBeInTheDocument();
  });

  it("should show an error message when there is no data (parsing issue 1/2)", () => {
    const { getByTestId, getByText } = render(
      <ComparisonTable
        newNode={{ ...baseNode, props: "NOT JSON" }}
        existingNode={{ ...baseNode, props: "{}" }}
        loading={false}
      />
    );

    expect(getByTestId(/node-comparison-error/i)).toBeInTheDocument();
    expect(getByText(/Unable to show comparison of data/i)).toBeInTheDocument();
  });

  it("should show an error message when there is no data (parsing issue 2/2)", () => {
    const { getByTestId, getByText } = render(
      <ComparisonTable
        newNode={{ ...baseNode, props: "{}" }}
        existingNode={{ ...baseNode, props: "NOT JSON" }}
        loading={false}
      />
    );

    expect(getByTestId(/node-comparison-error/i)).toBeInTheDocument();
    expect(getByText(/Unable to show comparison of data/i)).toBeInTheDocument();
  });

  it("should show a loading skeleton when loading", () => {
    const { getAllByTestId } = render(
      <ComparisonTable newNode={null} existingNode={null} loading />
    );

    expect(getAllByTestId("node-comparison-table-header-skeleton")).toHaveLength(5);
  });

  it("should rerender if a dependency changes", () => {
    const { rerender, getByText, queryByText } = render(
      <ComparisonTable
        newNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 01", baz: 1 }),
        }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 02", baz: 2 }),
        }}
        loading={false}
      />
    );

    expect(getByText(/example 01/i)).toBeInTheDocument();
    expect(getByText(/example 02/i)).toBeInTheDocument();

    rerender(
      <ComparisonTable
        newNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 03", baz: 1 }),
        }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 04", baz: 2 }),
        }}
        loading={false}
      />
    );

    expect(queryByText(/example 01/i)).not.toBeInTheDocument();
    expect(queryByText(/example 02/i)).not.toBeInTheDocument();
    expect(getByText(/example 03/i)).toBeInTheDocument();
    expect(getByText(/example 04/i)).toBeInTheDocument();
  });
});

describe("Implementation Requirements", () => {
  it("should show a table with the correct number of columns", () => {
    const { getByTestId } = render(
      <ComparisonTable
        newNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 01", baz: 1 }),
        }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify({ mock_node_data_name: "example 02", baz: 2 }),
        }}
        loading={false}
      />
    );

    expect(getByTestId("node-comparison-table")).toBeInTheDocument();
    expect(getByTestId("node-comparison-table").querySelectorAll("th")).toHaveLength(2);
  });

  it("should contain all of the header cells for each property", () => {
    const dataset = {
      mock_node_data_name: "example 01",
      another_property: "example 02",
      "property.with.dots": "yes",
    };

    const { getByText } = render(
      <ComparisonTable
        newNode={{
          ...baseNode,
          props: JSON.stringify(dataset),
        }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify(dataset),
        }}
        loading={false}
      />
    );

    expect(getByText(/mock_node_data_name/i)).toBeInTheDocument();
    expect(getByText(/another_property/i)).toBeInTheDocument();
    expect(getByText(/property.with.dots/i)).toBeInTheDocument();
  });

  // NOTE: this should never happen, but covering for completeness
  it("should aggregate all of the properties from both nodes", () => {
    const { getByText } = render(
      <ComparisonTable
        newNode={{
          ...baseNode,
          props: JSON.stringify({ mock_data_01: "example 01", another_md_02: 1 }),
        }}
        existingNode={{
          ...baseNode,
          props: JSON.stringify({ mock_data_02: "example 02", another_md_03: 2 }),
        }}
        loading={false}
      />
    );

    // new
    expect(getByText(/mock_data_01/i)).toBeInTheDocument();
    expect(getByText(/another_md_02/i)).toBeInTheDocument();

    // existing
    expect(getByText(/mock_data_02/i)).toBeInTheDocument();
    expect(getByText(/another_md_03/i)).toBeInTheDocument();
  });
});

describe("Snapshots", () => {
  it("should match the loading state snapshot", () => {
    const { container } = render(<ComparisonTable newNode={null} existingNode={null} loading />);

    expect(container).toMatchSnapshot("loading table");
  });
});
