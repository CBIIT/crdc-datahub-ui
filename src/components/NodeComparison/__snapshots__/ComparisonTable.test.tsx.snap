// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should match the loading state snapshot: loading table 1`] = `
<div>
  <div
    class="MuiTableContainer-root css-1dgj49d-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-1sjtdyy-MuiTable-root"
      data-testid="node-comparison-table"
    >
      <thead
        class="MuiTableHead-root css-d1l5r2-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            width="55"
          />
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            data-testid="node-comparison-table-header-skeleton"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            data-testid="node-comparison-table-header-skeleton"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            data-testid="node-comparison-table-header-skeleton"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            data-testid="node-comparison-table-header-skeleton"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-14un85q-MuiTableCell-root"
            data-testid="node-comparison-table-header-skeleton"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
            width="55"
          >
            Existing
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
            width="55"
          >
            New
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1w8d13p-MuiTableCell-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              style="width: 100%; height: 24px;"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
